Mathlib/Data/Set/Subset.lean

```diff
@@ -0,0 +1,346 @@
+  exact h
+
+lemma restrict_mono (h : B ⊆ C) : A ↓∩ B ⊆ A ↓∩ C := by
+  simp only [restrict_subset_restrict_iff, subset_inter_iff, inter_subset_left, true_and]
+  apply subset_trans (inter_subset_right A B) h
+
+lemma mem_coe_iff (x : α): x ∈  (↑D : Set α) ↔ ∃ y : ↑A, y ∈ D ∧ ↑y = x  := by rfl
+
+/--
+The following simp lemmas try to transform operations in the subtype into operations  in the ambient
+type, if possible.
+-/
+
+lemma coe_univ : ↑(univ : Set A)  = A := by
+  simp only [image_univ, Subtype.range_coe_subtype, setOf_mem_eq]
+
+lemma coe_empty : ↑(∅ : Set A)  = (∅ : Set α ) := image_empty _
+
+@[simp]
+lemma coe_union : (↑(D ∪ E) : Set α)  = ↑D ∪ ↑E := by
+  ext x
+  simp_all only [mem_union, mem_image, Subtype.exists, exists_and_right, exists_eq_right]
+  apply Iff.intro
+  · intro a
+    aesop_destruct_products
+    simp_all only [exists_true_left]
+  · intro a
+    cases a
+    · aesop_destruct_products
+      simp_all only [true_or, exists_const]
+    · aesop_destruct_products
+      simp_all only [or_true, exists_const]
+
+@[simp]
+lemma coe_inter : (↑(D ∩ E) : Set α) = ↑D  ∩ ↑E := by
+  ext
+  simp_all only [mem_inter_iff, mem_image, Subtype.exists, exists_and_right, exists_eq_right]
+  apply Iff.intro
+  · intro a
+    aesop_destruct_products
+    simp_all only [and_self, exists_const]
+  · intro a
+    aesop_destruct_products
+    simp_all only [exists_const, and_self]
+
+@[simp]
+lemma coe_compl : ↑(Dᶜ) = A \ ↑D := by
+  unhygienic ext
+  simp_all only [mem_image, mem_compl_iff, Subtype.exists, exists_and_right, exists_eq_right,
+    mem_diff, not_exists]
+  apply Iff.intro
+  · intro a
+    cases a
+    simp_all only [not_false_eq_true, forall_true_left, and_self]
+  · intro a
+    simp_all only [not_false_eq_true, exists_const]
+
+@[simp]
+lemma coe_diff :  (↑(D \ E) : Set α) =  ↑D \ ↑E  := by
+  ext
+  simp_all only [mem_diff, mem_image, Subtype.exists, exists_and_right, exists_eq_right, not_exists]
+  apply Iff.intro
+  · intro a
+    aesop_destruct_products
+    simp_all only [exists_const, not_false_eq_true, forall_true_left, and_self]
+  · intro a
+    simp_all only [not_false_eq_true, and_true]
+
+@[simp]
+lemma coe_sUnion : ↑(⋃₀ T)  = ⋃₀ { (B : Set α) | B ∈ T} := by
+  ext x
+  simp_all only [mem_sUnion, mem_setOf_eq, exists_exists_and_eq_and, mem_image, Subtype.exists,
+    exists_and_right, exists_eq_right]
+  apply Iff.intro
+  · intro a
+    unhygienic with_reducible aesop_destruct_products
+    simp_all only [exists_true_left]
+    apply Exists.intro
+    apply And.intro
+    on_goal 2 => exact right
+    simp_all only
+  · intro a
+    unhygienic with_reducible aesop_destruct_products
+    simp_all only [exists_true_left]
+    apply Exists.intro
+    apply And.intro
+    exact left
+    simp_all only
+
+
+@[simp]
+lemma coe_iUnion : ↑(⋃ (B : β ), j B) = ⋃ (B : β), (j B : Set α) := image_iUnion
+
+@[simp]
+lemma coe_sInter (hT : ∃ L, L ∈ T) : (↑(⋂₀ T) : Set α) = ⋂₀ { (↑B : Set α) | B ∈ T}  := by
```

Comment:
Use `Set.Nonempty` so its associated API can be used.
```suggestion
lemma coe_sInter (hT : T.Nonempty) : (↑(⋂₀ T) : Set α) = ⋂₀ { (↑B : Set α) | B ∈ T}  := by
```

Reply:
Done.